quickwit/quickwit-indexing/src/source/queue_sources/message.rs

```diff
@@ -0,0 +1,295 @@
+use crate::source::doc_file_reader::ObjectUriBatchReader;
+
+#[derive(Debug, Clone, Copy)]
+pub enum MessageType {
+    S3Notification,
+    // GcsNotification,
+    RawUri,
+    // RawData,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct MessageMetadata {
+    /// The handle that should be used to acknowledge the message or change its visibility deadline
+    pub ack_id: String,
+
+    /// The unique message id assigned by the queue
+    pub message_id: String,
+
+    /// The approximate number of times the message was delivered. 1 means it is
+    /// the first time this message is being delivered.
+    pub delivery_attempts: usize,
+
+    /// The first deadline when the message is received. It can be extended later using the ack_id.
+    pub initial_deadline: Instant,
+}
+
+/// The raw messages as received from the queue abstraction
+pub struct RawMessage {
+    pub metadata: MessageMetadata,
+    pub payload: OwnedBytes,
+}
+
+impl fmt::Debug for RawMessage {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("RawMessage")
+            .field("metadata", &self.metadata)
+            .field("payload", &"<bytes>")
+            .finish()
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum PreProcessingError {
+    /// The message can be acknowledged without processing
+    #[error("skippable message: {reason}")]
+    Discardable {
+        reason: &'static str,
+        ack_id: String,
+    },
+    #[error("unexpected message format: {0}")]
+    UnexpectedFormat(#[from] anyhow::Error),
+}
+
+impl RawMessage {
+    pub fn pre_process(
+        self,
+        message_type: MessageType,
+    ) -> Result<PreProcessedMessage, PreProcessingError> {
+        let payload = match message_type {
+            MessageType::S3Notification => PreProcessedPayload::ObjectUri(
+                uri_from_s3_notification(&self.payload, &self.metadata.ack_id)?,
+            ),
+            MessageType::RawUri => {
+                let payload_str = read_to_string(self.payload).context("failed to read payload")?;
+                PreProcessedPayload::ObjectUri(Uri::from_str(&payload_str)?)
+            }
+        };
+        Ok(PreProcessedMessage {
+            metadata: self.metadata,
+            payload,
+        })
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum PreProcessedPayload {
+    ObjectUri(Uri),
+}
+
+impl PreProcessedPayload {
+    pub fn partition_id(&self) -> PartitionId {
+        match &self {
+            Self::ObjectUri(uri) => PartitionId::from(uri.as_str()),
+        }
+    }
+}
+
+/// A message that went through the minimal transformation to discover its
+/// partition id. Indeed, the message might be discarded if the partition was
+/// already processed, so it's better to avoid doing unnecessary work at this
+/// stage.
+#[derive(Debug, PartialEq, Eq)]
+pub struct PreProcessedMessage {
+    pub metadata: MessageMetadata,
+    pub payload: PreProcessedPayload,
```

Comment:
payload is too generic. What is this payload about?
Right now I'd rather have a simple `Uri`

Reply:
I agree that we can apply YAGNI here. But at the same time, I would like to quickly use this to re-implement the Google Pub/Sub source in a follow up PR.